### PR TITLE
Pending Processes

### PR DESCRIPTION
**Pull request by nzoschke**
```
$ convox ps --app httpd
ID            NAME  RELEASE      SIZE  STARTED       COMMAND         
8d14ced88808  web   RRGNOZOTWGB  768   20 hours ago  httpd-foreground
e27d32c8bdf7  web   RRGNOZOTWGB  768   20 hours ago  httpd-foreground
[PENDING]     web   RQICSWKWYRC  768                                 
[PENDING]     web   RQICSWKWYRC  768  

$ convox scale --app httpd
NAME  DESIRED  RUNNING  MEMORY
web   2        2        768
```

### CODE REVIEW

**Comment by nzoschke** [comment]
This should probably split the Docker stuff out into it's own test server and cycles. This will probably help with the concurrency challenges here.

**Comment by csquared** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;q7F9Og2j&#x2F;245-surface-insufficient-capacity-errors"> surface insufficient capacity errors</a>

File: api/models/process.go [hunk 1]
```diff
@@ -22,6 +23,7 @@ type Process struct {
 	Name    string    `json:"name"`
 	Ports   []string  `json:"ports"`
 	Release string    `json:"release"`
+	Size    int64     `json:"size"`
```
Comment (csquared) [inline api/models/process.go]: I wonder if there is name that could more strongly hint this is the size of memory and not cpu shares or something else. `MemSize` ? `RequestedMemory` ? `ProvisionedMemory` ? Just some food for thought.

File: api/models/process.go [hunk 2]
```diff
@@ -114,11 +116,81 @@ func ListProcesses(app string) (Processes, error) {
...
+			continue
+		}
+
+		// Test every service deployment for running != pending, to put in a placeholder
+		for _, d := range service.Deployments {
+			if *d.DesiredCount == *d.RunningCount {
+				continue
+			}
+
+			tres, err := ECS().DescribeTaskDefinition(&ecs.DescribeTaskDefinitionInput{
```
Comment (csquared) [inline api/models/process.go]: is this api request happening serially?  will that scale OK for bigger apps?

File: api/models/process.go [hunk 3]
```diff
@@ -178,12 +251,17 @@ func fetchProcess(app string, task ecs.Task, td ecs.TaskDefinition, cd ecs.Conta
...
 	if len(ires.Reservations) != 1 || len(ires.Reservations[0].Instances) != 1 {
-		errch <- fmt.Errorf("could not find instance")
+		errch <- fmt.Errorf("could not find EC2 instance")
 		return
 	}
 
 	instance := ires.Reservations[0].Instances[0]
 
+	// Connect to a Docker client
+	// In testing use the stub AWS server.
```
Comment (csquared) [inline api/models/process.go]: not the AWS server anymore, right?

File: api/models/process.go [hunk 4]
```diff
@@ -266,6 +357,10 @@ func (p *Process) FetchStats() error {
 	d, err := p.Docker()
 
 	if err != nil {
+		if err.Error() == "pending" {
```
Comment (csquared) [inline api/models/process.go]: is it safer to make an error type and check for that?

File: test/aws_cycles.go [hunk 5]
```diff
@@ -150,6 +149,374 @@ func ListContainerInstancesCycle(clusterName string) awsutil.Cycle {
...
+func DescribeServicesCycle(clusterName string) awsutil.Cycle {
+	return awsutil.Cycle{
+		Request: awsutil.Request{
+			RequestURI: "/",
+			Operation:  "AmazonEC2ContainerServiceV20141113.DescribeServices",
+			Body:       `{"cluster":"` + clusterName + `", "services":["arn:aws:ecs:us-west-2:901416387788:service/myapp-staging-worker-SCELGCIYSKF"]}`,
+		},
+		Response: awsutil.Response{
+			StatusCode: 200,
+			Body: `
```
Comment (csquared) [inline test/aws_cycles.go]: I'd get this body in its own method so the cycle is a little clearer

File: test/aws_cycles.go [hunk 6]
```diff
@@ -150,6 +149,374 @@ func ListContainerInstancesCycle(clusterName string) awsutil.Cycle {
...
+func DescribeServicesWithDeploymentsCycle(clusterName string) awsutil.Cycle {
+	return awsutil.Cycle{
+		Request: awsutil.Request{
+			RequestURI: "/",
+			Operation:  "AmazonEC2ContainerServiceV20141113.DescribeServices",
+			Body:       `{"cluster":"` + clusterName + `", "services":["arn:aws:ecs:us-west-2:901416387788:service/myapp-staging-worker-SCELGCIYSKF"]}`,
+		},
+		Response: awsutil.Response{
+			StatusCode: 200,
+			Body: `
```
Comment (csquared) [inline test/aws_cycles.go]: same for this -> would just tuck the body away

File: test/aws_cycles.go [hunk 7]
```diff
@@ -150,6 +149,374 @@ func ListContainerInstancesCycle(clusterName string) awsutil.Cycle {
...
+],"failures":[]}`},
+	}
+}
+
+func DescribeInstancesFilteredCycle() awsutil.Cycle {
+	return awsutil.Cycle{
+		awsutil.Request{"/", "", `Action=DescribeInstances&Filter.1.Name=instance-id&Filter.1.Value.1=i-4a5513f4&Version=2015-10-01`},
+		awsutil.Response{200, `<DescribeInstancesResponse xmlns="http://ec2.amazonaws.com/doc/2015-10-01/">
```
Comment (csquared) [inline test/aws_cycles.go]: esp. this response

File: test/docker_cycles.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+func StatsCycle() awsutil.Cycle {
+	return awsutil.Cycle{
+		Request: awsutil.Request{
+			RequestURI: "/containers/8dfafdbc3a40/stats?stream=false",
+			Operation:  "",
+			Body:       ``,
+		},
+		Response: awsutil.Response{
+			StatusCode: 200,
+			Body: `{
```
Comment (csquared) [inline test/docker_cycles.go]: last response body to tuck away

**Comment by csquared** [comment]
functionality is super awesome. would like to see the tests cleaned up (by putting stuff in methods) before merging, i think everything else is just a suggestion

File: api/models/process.go (hunk 3 above)
Comment (nzoschke) [inline api/models/process.go]: fixed

**Pull request merged**